KubectlApply doesn't work with some Rancher clusters.

**Describe the bug**
  I am trying to create an service using Kubectl.apply as per this: https://github.com/kubernetes-client/java/blob/b0c91798c63203f20dc8f2674197e772b2c187ee/docs/kubectl-equivalence-in-java.md#kubectl-apply

  source code:
  ```
  KubectlApply kubectlApply = Kubectl.apply(V1Service.class).fieldManager("kubectl").resource(resource);
  KubectlApply kubectlApplyClient = kubectlApply.apiClient(apiClient);
  kubectlApplyClient.execute();
  ```

  I am getting below error:

  ```
      Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Not a JSON Object: 404
      at com.google.gson.Gson.fromJson(Gson.java:944)
      at com.google.gson.Gson.fromJson(Gson.java:897)
      at com.google.gson.Gson.fromJson(Gson.java:846)
      at io.kubernetes.client.openapi.JSON.deserialize(JSON.java:169)
      at io.kubernetes.client.util.generic.GenericKubernetesApi.patch(GenericKubernetesApi.java:632)
      at io.kubernetes.client.extended.kubectl.KubectlApply.executeServerSideApply(KubectlApply.java:89)
      at io.kubernetes.client.extended.kubectl.KubectlApply.execute(KubectlApply.java:67)
      at chengf.falcon.k8s.learn.ManifestTest.applyManifest(ManifestTest.java:65)
      ... 23 more
      Caused by: java.lang.IllegalStateException: Not a JSON Object: 404
      at com.google.gson.JsonElement.getAsJsonObject(JsonElement.java:91)
      at io.kubernetes.client.gson.V1StatusPreProcessor.isLegitKubernetesStatus(V1StatusPreProcessor.java:39)
      at io.kubernetes.client.gson.V1StatusPreProcessor.preDeserialize(V1StatusPreProcessor.java:26)
      at io.gsonfire.gson.HooksTypeAdapter.runPreDeserialize(HooksTypeAdapter.java:79)
      at io.gsonfire.gson.HooksTypeAdapter.read(HooksTypeAdapter.java:53)
      at com.google.gson.Gson.fromJson(Gson.java:932)
      ... 30 more
  ```

**Client Version**
I've tried the following version：

`11.0.0`、`11.0.2`、`13.0.0`

**Kubernetes Version**
`v1.19.0`

**Java Version**
Java 8

**service.yaml**
```
apiVersion: v1
kind: Service
metadata:
  name: nginx-test
  namespace: appstore
  labels:
    app: nginx-test
spec:
  ports:
    - port: 80
      name: http
      targetPort: 80
      protocol: TCP
  selector:
    app: nginx-test

```

I tried the same resource with kubectl CLI and I don't see any error:
```
kubectl apply --server-side=true --field-manager=kubectl -f service.yaml
```

@dddddai Sorry for the late reply
What client version are you using. As I tried the same code you shared, the error is still there

* Source code
```
{
        ApiClient apiClient = ClientBuilder.kubeconfig(KubeConfig.loadKubeConfig(new FileReader(Thread.currentThread().getContextClassLoader().getResource("config.yml").getPath()))).build();
        Configuration.setDefaultApiClient(apiClient);
        V1ServiceSpec spec=new V1ServiceSpec();
        spec.setSelector(Collections.singletonMap("app","foo"));
        V1ServicePort port = new V1ServicePort();
        port.setPort(8440);
        port.setTargetPort(new IntOrString(80));
        port.setName("http");
        port.setProtocol("TCP");
        spec.setPorts(Collections.singletonList(port));

        V1Service service =
                Kubectl.apply(V1Service.class)
                        .resource(new V1Service()
                                .kind("Service")
                                .apiVersion("v1")
                                .metadata(new V1ObjectMeta().namespace("dwq").name("foo").labels(Collections.singletonMap("app","foo")))
                                .spec(spec))
                        .execute();
    }
```

* Error message
```
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Not a JSON Object: 404

	at com.google.gson.Gson.fromJson(Gson.java:944)
	at com.google.gson.Gson.fromJson(Gson.java:897)
	at com.google.gson.Gson.fromJson(Gson.java:846)
	at io.kubernetes.client.openapi.JSON.deserialize(JSON.java:169)
	at io.kubernetes.client.util.generic.GenericKubernetesApi.patch(GenericKubernetesApi.java:632)
	at io.kubernetes.client.extended.kubectl.KubectlApply.executeServerSideApply(KubectlApply.java:89)
	at io.kubernetes.client.extended.kubectl.KubectlApply.execute(KubectlApply.java:67)
	at chengf.falcon.k8s.learn.KubectlApplyServiceTest.test2(KubectlApplyServiceTest.java:104)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: java.lang.IllegalStateException: Not a JSON Object: 404
	at com.google.gson.JsonElement.getAsJsonObject(JsonElement.java:91)
	at io.kubernetes.client.gson.V1StatusPreProcessor.isLegitKubernetesStatus(V1StatusPreProcessor.java:39)
	at io.kubernetes.client.gson.V1StatusPreProcessor.preDeserialize(V1StatusPreProcessor.java:26)
	at io.gsonfire.gson.HooksTypeAdapter.runPreDeserialize(HooksTypeAdapter.java:79)
	at io.gsonfire.gson.HooksTypeAdapter.read(HooksTypeAdapter.java:53)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	... 29 more
```

* Client version

```
<dependency>
            <groupId>io.kubernetes</groupId>
            <artifactId>client-java</artifactId>
            <version>13.0.0</version>
        </dependency>
        <dependency>
            <groupId>io.kubernetes</groupId>
            <artifactId>client-java-extended</artifactId>
            <version>13.0.0</version>
        </dependency>


        <dependency>
            <groupId>io.kubernetes</groupId>
            <artifactId>client-java-api-fluent</artifactId>
            <version>13.0.0</version>
        </dependency>
```

@dddddai  ```Kubectl.get``` command works. 

I tried ```Kubectl.apply``` command for ```Deployment```,```DaemonSet```,```ClusterRole```, all works fine. 

```Kubectl.create``` for ```Service``` throws similar error

 ```
io.kubernetes.client.extended.kubectl.exception.KubectlException: io.kubernetes.client.openapi.ApiException: class V1Status {
    apiVersion: null
    code: 404
    details: null
    kind: null
    message: 404 page not found
    
    metadata: null
    reason: null
    status: null
}

	at io.kubernetes.client.extended.kubectl.KubectlCreate.execute(KubectlCreate.java:58)
	at chengf.falcon.k8s.learn.KubectlApplyServiceTest.test2(KubectlApplyServiceTest.java:104)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: io.kubernetes.client.openapi.ApiException: class V1Status {
    apiVersion: null
    code: 404
    details: null
    kind: null
    message: 404 page not found
    
    metadata: null
    reason: null
    status: null
}
	at io.kubernetes.client.util.generic.KubernetesApiResponse.lambda$throwsApiException$0(KubernetesApiResponse.java:64)
	at io.kubernetes.client.util.generic.KubernetesApiResponse.onFailure(KubernetesApiResponse.java:78)
	at io.kubernetes.client.util.generic.KubernetesApiResponse.throwsApiException(KubernetesApiResponse.java:62)
	at io.kubernetes.client.extended.kubectl.KubectlCreate.execute(KubectlCreate.java:55)
	... 23 more

```

--- a/extended/src/test/java/io/kubernetes/client/extended/kubectl/KubectlApplyTest.java
+++ b/extended/src/test/java/io/kubernetes/client/extended/kubectl/KubectlApplyTest.java
@@ -12,9 +12,12 @@ limitations under the License.
 */
 package io.kubernetes.client.extended.kubectl;
 
-import static com.github.tomakehurst.wiremock.client.WireMock.*;
+import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
+import static com.github.tomakehurst.wiremock.client.WireMock.get;
+import static com.github.tomakehurst.wiremock.client.WireMock.patch;
+import static com.github.tomakehurst.wiremock.client.WireMock.patchRequestedFor;
 import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
-import static org.junit.Assert.*;
+import static org.junit.Assert.assertNotNull;
 
 import com.github.tomakehurst.wiremock.junit.WireMockRule;
 import com.github.tomakehurst.wiremock.matching.EqualToPattern;

--- a/util/src/main/java/io/kubernetes/client/util/generic/GenericKubernetesApi.java
+++ b/util/src/main/java/io/kubernetes/client/util/generic/GenericKubernetesApi.java
@@ -824,7 +824,17 @@ public class GenericKubernetesApi<
       return call;
     }
     HttpUrl url = call.request().url();
-    HttpUrl tweakedUrl = url.newBuilder().removePathSegment(1).setPathSegment(0, "api").build();
+    String basePath = customObjectsApi.getApiClient().getBasePath();
+    // count segments in the basePath
+    int offset = 0;
+    for (int i = basePath.indexOf("://") + 3; i < basePath.length(); ++i) {
+      if (basePath.charAt(i) == '/') {
+        ++offset;
+      }
+    }
+    HttpUrl tweakedUrl =
+        url.newBuilder().removePathSegment(offset + 1).setPathSegment(offset, "api").build();
+
     return this.customObjectsApi
         .getApiClient()
         .getHttpClient()

--- a/util/src/test/java/io/kubernetes/client/util/generic/GenericKubernetesApiForCoreApiTest.java
+++ b/util/src/test/java/io/kubernetes/client/util/generic/GenericKubernetesApiForCoreApiTest.java
@@ -170,6 +170,36 @@ public class GenericKubernetesApiForCoreApiTest {
   }
 
   @Test
+  public void patchNamespacedPodWithApiPrefix() {
+    V1Patch v1Patch = new V1Patch("{}");
+    V1Pod foo1 =
+        new V1Pod().kind("Pod").metadata(new V1ObjectMeta().namespace("default").name("foo1"));
+    // add api prefix
+    String prefix = "/k8s/clusters/c-7q988";
+    stubFor(
+        patch(urlEqualTo(prefix + "/api/v1/namespaces/default/pods/foo1"))
+            .withHeader("Content-Type", containing(V1Patch.PATCH_FORMAT_STRATEGIC_MERGE_PATCH))
+            .willReturn(aResponse().withStatus(200).withBody(new Gson().toJson(foo1))));
+
+    GenericKubernetesApi<V1Pod, V1PodList> rancherPodClient =
+        new GenericKubernetesApi<>(
+            V1Pod.class,
+            V1PodList.class,
+            "",
+            "v1",
+            "pods",
+            new ClientBuilder().setBasePath("http://localhost:" + 8181 + prefix).build());
+    KubernetesApiResponse<V1Pod> podPatchResp =
+        rancherPodClient.patch(
+            "default", "foo1", V1Patch.PATCH_FORMAT_STRATEGIC_MERGE_PATCH, v1Patch);
+
+    assertTrue(podPatchResp.isSuccess());
+    assertEquals(foo1, podPatchResp.getObject());
+    assertNull(podPatchResp.getStatus());
+    verify(1, patchRequestedFor(urlPathEqualTo(prefix + "/api/v1/namespaces/default/pods/foo1")));
+  }
+
+  @Test
   public void testReadTimeoutShouldThrowException() {
     ApiClient apiClient = new ClientBuilder().setBasePath("http://localhost:" + 8181).build();
     apiClient.setHttpClient(

